lib/iris/experimental/ugrid/mesh.py

```diff
@@ -975,77 +925,129 @@ def __ne__(self, other):
+        if mesh_name:
+            # Use a more human-readable form
+            mesh_string = f"<Mesh: '{mesh_name}'>"
+        else:
+            # Mimic the generic object.__str__ style.
+            mesh_id = id(self)
+            mesh_string = f"<Mesh object at {hex(mesh_id)}>"
+
+        return mesh_string
+
+    def _summary_multiline(self):
+        # Produce a readable multi-line summary of the Mesh content.
+        lines = []
+        n_indent = 4
+        indent_str = " " * n_indent
+
+        def line(text, i_indent=0):
+            indent = indent_str * i_indent
+            lines.append(f"{indent}{text}")
+
+        line(f"Mesh : '{self.name()}'")
+        line(f"topology_dimension: {self.topology_dimension}", 1)
+        for element in ("node", "edge", "face"):
+            if element == "node":
+                element_exists = True
+            else:
+                main_conn_name = f"{element}_node_connectivity"
+                main_conn = getattr(self, main_conn_name, None)
+                element_exists = main_conn is not None
+            if element_exists:
+                # Include a section for this element
+                line(element, 1)
+                # Print element dimension
+                dim_name = f"{element}_dimension"
+                dim = getattr(self, dim_name)
+                line(f"{dim_name}: '{dim}'", 2)
+                # Print defining connectivity (except node)
+                if element != "node":
+                    main_conn_string = main_conn.summary(
+                        shorten=True, linewidth=0
+                    )
+                    line(f"{main_conn_name}: {main_conn_string}", 2)
+                # Print coords
+                include_key = f"include_{element}s"
+                coords = self.coords(**{include_key: True})
+                if coords:
+                    line(f"{element} coordinates", 2)
+                    for coord in coords:
+                        coord_string = coord.summary(shorten=True, linewidth=0)
+                        line(coord_string, 3)
+
+        # Having dealt with essential info, now add any optional connectivities
+        # N.B. includes boundaries: as optional connectivity, not an "element"
+        optional_conn_names = (
+            "boundary_connectivity",
+            "face_face_connectivity",
+            "face_edge_connectivity",
+            "edge_face_connectivity",
         )
+        optional_conns = [
+            getattr(self, name, None) for name in optional_conn_names
+        ]
+        optional_conns = {
+            name: conn
+            for conn, name in zip(optional_conns, optional_conn_names)
+            if conn is not None
+        }
+        if optional_conns:
+            line("optional connectivities", 1)
+            for name, conn in optional_conns.items():
+                conn_string = conn.summary(shorten=True, linewidth=0)
+                line(f"{name}: {conn_string}", 2)
```

Comment:
Could there be a case for grouping these optional connectivities together by their location/source element. e.g.

edge:
`edge_node_connectivity`
(optionally)
`edge_face_connectivity`

face:
`face_node_connectivity`
(optionally)
`face_edge_connectivity`
`face_face_connectivity`

with `boundary_node_connectivity` as a special case on its own.
In this way, connectivities would be grouped with other connectivities which share their dimensions.
(As an aside, I don't believe we currently keep hold of the NetCDF dimension associated with `boundary_node_connectivity`, I'm not sure if this is something we should be concerned about. I had a look at what UGRID says about them and it seems to suggest that `boundary_node_connectivity` can have ancillary variables attached which I don't know if we support. This is probably not core functionality, but it might be worth recording as an issue for later.)
With that said, I could also imagine an argument that something like `edge_face_connectivity` is just as much assiciated with faces as it is edges. I'm not necessarily asking for a change, but I thought it would be worth at least raising the point for discussion.

Reply:
At present I am purposely listing the _non_-optional 'edge_node' and 'face_node' connectivities in their "element section", as you suggest.

In my view, the "essential" connectivities really have a different scope to the 'optional' ones, and should be treated differently.  
I think If we put the 'optional' ones into their relevant sections, that's maybe not so clear.
In that case I think I would to list the required information first, and add the optional connectivities after the coords (if any).  Whether they need to be labelled 'optional' I'm not quite so sure.

As examples from an actual file ...
CURRENT
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
    optional connectivities
        face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
        face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
        edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

OPTIONALS IN OWN SECTIONS
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
        face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

OPTIONALS LABELLED AS SUCH
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        optional connectivities:
            edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        optional connectivities:
            face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
            face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

WITH BOUNDARIES
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        optional connectivities:
            edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        optional connectivities:
            face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
            face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    boundaries
        boundary_dimension: 'nMesh2d_half_levels_boundary'
        boundary_node_connectivity: <Connectivity: unknown / (unknown)  <lazy>  shape(1417, 2)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

Reply:
(After all that wordage)
I conclude that its perfectly possible, and it might make sense to put boundaries in their own section, and report the 'boundary_dimension', like this, but in that case I think we should support it in the Mesh object too.

But i think there are some CONs too : 
  * the existing output is more concise 
  * the existing output is possibly slightly clearer
  * I'm not convinced that we really know what the "status" of boundaries is : it seems to be a feature that is not fully developed + it's not clear quite how it is intended to be used or what for

Reply:
I agree that this is definitely a worthwhile thing to discuss. The hard part is knowing what the best option is! I'm strugglling to come to a conclusion on this. I'm not terribly familiar with the area myself so finding it hard to decide what a user would expect. However, I am leaning towards keeping things as they are , primarily because I like the conciseness of the original output, also @stephenworsley's point:
> With that said, I could also imagine an argument that something like edge_face_connectivity is just as much assiciated with faces as it is edges. 
This is an interesting point! 

I think it might be simpler to keep the optional connectivities in their own section for now. But readdress this point at a later stage. Would you be okay with that @stephenworsley ?

Reply:
Yeah, I think either way is workable and keeping it as it is currently is probably for the best for the moment. I imagine this could be something we could return to when we have a better idea of what to do with the boundaries.